### PR TITLE
Pg/bases

### PR DESCRIPTION
**Pull request by pgawron**
New bases system and major clean-up.

### CODE REVIEW

File: src/base.jl [hunk 1]
```diff
@@ -1,21 +1,34 @@
...
     1 <= val <= dim ? () : throw(ArgumentError("Label have to be smaller than vector dimension"))
-    ψ = zeros(T, dim)
-    ψ[val] = one(T)
+    ψ = zero(T(undef, dim))
+    ψ[val] = one(eltype(T))
     ψ
 end
 
+function ket(::Type{T}, val::Int, dim::Int) where T<:Number
+    @warn "This method is deprecated and will be removed. Use calls like `ket(Matrix{ComplexF64}, 1, 2)`."
```
Comment (lpawela) [inline src/base.jl]: Maybe we should keep both?

File: src/base.jl [hunk 2]
```diff
@@ -24,16 +37,21 @@ $(SIGNATURES)
...
-bra(val::Int, dim::Int) = bra(ComplexF64, val, dim)
+bra(val::Int, dim::Int) = bra(Vector{ComplexF64}, val, dim)
 
-function ketbra(::Type{T}, valk::Int, valb::Int, dim::Int) where T<:Number
+function ketbra(::Type{T}, valk::Int, valb::Int, dim::Int) where T<:AbstractMatrix{<:Number}
     dim > 0 ? () : throw(ArgumentError("Vector dimension has to be nonnegative"))
     1 <= valk <= dim && 1 <= valb <= dim ? () : throw(ArgumentError("Ket and bra labels have to be smaller than operator dimension"))
-    ρ = zeros(T, dim, dim)
-    ρ[valk,valb] = one(T)
+    ρ = zero(T(undef, dim, dim))
```
Comment (lpawela) [inline src/base.jl]: We should support the case of rectangular matrices. Add idim (bra dim), odim (ket dim).

File: src/gates.jl [hunk 3]
```diff
@@ -1,3 +1,10 @@
+export 𝕀, sx, sy, sz, qft, hadamard, grover
+
+sx = ComplexF64[0 1; 1 0]
+sy = ComplexF64[0 1im; -1im 0]
+sz = ComplexF64[1 0; 0 -1]
+𝕀(::Type{T}, dim=2) where T<: Number = Matrix{T}(I, dim, dim)
```
Comment (lpawela) [inline src/gates.jl]: This should be deprecated in new `julia` versions. `I` works

File: src/base.jl (hunk 1 above)
Comment (pgawron) [reply src/base.jl]: Too much clutter in my opinion.

File: src/base.jl (hunk 2 above)
Comment (pgawron) [reply src/base.jl]: Fixed

File: src/gates.jl (hunk 3 above)
Comment (pgawron) [reply src/gates.jl]: Fixed

**Pull request merged**